cpu/nrf52: add PWM implementation

SemjonKerner opened the pull request:
This is an initial peripheral pwm implementation for nordic nrf52 cpu. This implementation provides all functions in pwm.h. It implements nordics hardware pwm. Also this PR adds nrf52dk board configs.

PeterKietzmann commented on boards/common/nrf52xxxdk/include/periph_conf.h:
```diff
@@ -108,6 +108,17 @@ static const i2c_conf_t i2c_config[] = {
+ * @{
+ */
+static const pwm_conf_t pwm_config[] = {
+    { NRF_PWM0, { 23, 24, 25, 26 } }
```
P0.23 doesn't seem to work here. Does it on your board?

SemjonKerner replied:
@PeterKietzmann Yes, it works for me on nrf52dk. Is it possible, that you are testing a different board with P0.23 already in use by another module?

PeterKietzmann replied:
To be honest, I don't remember. **But** in my opinion any board that provides the PWM feature and makes use of this *periph_conf.h* should work with the provided configuration of **at least** be documented.

SemjonKerner replied:
I double check that, but i am confident 23 works for me, because i mostly used that pin for debugging.

PeterKietzmann replied:
Would be great if you could check the manuals of the other nrf52 boards that we support. I think it's just 2-3...

PeterKietzmann commented on cpu/nrf52/periph/pwm.c:
```diff
@@ -0,0 +1,184 @@
+
+    DEBUG("PWM started\n");
+
+    return real_clk;
```
I assume this value has to be adapted in case of  `PWM_CENTER` ?
```
init 0 1 10000 100
The pwm frequency is set to 10000

init 0 2 10000 100
The pwm frequency is set to 20000
```

However, frequency is 10k in both cases (as expected)

SemjonKerner replied:
@PeterKietzmann i intend on dropping center mode support. I already checked that there is no support in center mode for other CPUs, also center mode is rearly used anyway. If center mode has no special relevance to you, i will just drop it.

PeterKietzmann replied:
I'm okay with dropping that feature.

PeterKietzmann commented on cpu/nrf52/periph/pwm.c:
```diff
@@ -0,0 +1,184 @@
+    else {
+        pwm_seq[pwm][channel] = dev(pwm)->COUNTERTOP - value;
+    }
+
```
Can you please spend a note about the center mode here?

SemjonKerner commented:
@PeterKietzmann Okay, so i moved the pins and restricted the pwm pin configuration to both boards nrf52dk and nrf52840dk. I decided to use ifdefs, because it felt excessive to create new periph_conf files for both boards just to add 4 common pins each.
Also i dropped center mode.

PeterKietzmann commented on cpu/nrf52/include/periph_cpu.h:
```diff
@@ -108,6 +113,49 @@ typedef struct {
+typedef enum {
+    PWM_LEFT       = PWM_MODE(0, 1),    /**< left aligned PWM */
+    PWM_RIGHT      = PWM_MODE(0, 0),    /**< right aligned PWM */
+    PWM_CENTER     = PWM_MODE(1, 1),    /**< not supported */
```
Should we remove this?

SemjonKerner replied:
I don't have a strong opinion about that. As i saw in other cpus (samd21,efm32) center mode is not supported but PWM_CENTER is set anyway. I did not question this design, but i think it gives easy access / guidance to anyone who wants to implement center mode in the future.

PeterKietzmann replied:
Fine I leave it up to you. Please squash

PeterKietzmann commented:
`warning: Member PWM_NUMOF (macro definition) of file boards/common/nrf52xxxdk/include/periph_conf.h is not documented`

PeterKietzmann commented:
@SemjonKerner still around? Please fix Mudrocks complaints. My ACK holds

SemjonKerner commented:
@PeterKietzmann murdock is green
